org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java

```diff
@@ -811,58 +817,68 @@ public SSLEngineResult unwrap(ByteBuffer src, ByteBuffer[] dsts, int offset, int
 
-        // Maximum theoretical amount of data we could've written to the
-        // destination. This is bounded by the lower of both the size of
-        // our dsts and the maximum BUFFER_SIZE. Worst case, we'll be forced
-        // to call unwrap(...) multiple times.
-        int max_app_data = Math.min(computeSize(dsts, offset, length), BUFFER_SIZE);
-
-        // When we have app data to write over the network, go ahead and do
-        // so. This involves reading from ssl_fd and writing to dsts. We don't
-        // currently have a good proxy metric for "can read from a ssl_fd",
-        // so always attempt it. In particular, even if the handshake isn't
-        // finished, we still need to call PR.Read(...) or PR.Write(...) in
-        // order to tell if an inbound alert was received.
-        if (max_app_data > 0) {
-            byte[] app_buffer = PR.Read(ssl_fd, max_app_data);
-            debug("JSSEngine.unwrap() - " + app_buffer + " error=" + errorText(PR.GetError()));
+                    this_src_write = (int) Buffer.Write(read_buf, wire_buffer);
+
+                    wire_data += this_src_write;
+                    debug("JSSEngine.unwrap(): Wrote " + this_src_write + " bytes to read_buf.");
+                }
+            }
+
+            // In the above, we should always try to read and write data. Check to
+            // see if we need to step our handshake process or not.
+            updateHandshakeState();
+
+            int max_dst_size = computeSize(dsts, offset, length);
+            byte[] app_buffer = PR.Read(ssl_fd, max_dst_size);
+            int error = PR.GetError();
+            debug("JSSEngine.unwrap() - " + app_buffer + " error=" + errorText(error));
             if (app_buffer != null) {
-                app_data = putData(app_buffer, dsts, offset, length);
-            } else {
-                int error = PR.GetError();
-                if (error != 0 && error != PRErrors.WOULD_BLOCK_ERROR) {
+                this_dst_write = putData(app_buffer, dsts, offset, length);
+                app_data += this_dst_write;
+            } else if (max_dst_size > 0) {
+                // There are two scenarios we need to ignore here:
+                //  1. WOULD_BLOCK_ERRORs are safe, because we're expecting
+                //     not to block. Usually this means we don't have space
+                //     to write any more data.
+                //  2. SOCKET_SHUTDOWN_ERRORs are safe, because if the
+                //     underling cause was fatal, we'd catch it after exiting
+                //     the do-while loop, in checkSSLAlerts().
+                if (error != 0 && error != PRErrors.WOULD_BLOCK_ERROR && error != PRErrors.SOCKET_SHUTDOWN_ERROR) {
                     ssl_exception = new SSLException("Unexpected return from PR.Read(): " + errorText(error));
                     seen_exception = true;
                 }
             }
+        } while (this_src_write != 0 || this_dst_write != 0);
```

Comment:
I"m assuming that this_src_write and ths_dst_write will eventually become exactly zero?

Reply:
Yeah, I need to update the docs pr (#428) with this. But the theory is this: Let's take `wrap` as an example. We need to take data from `src` and put it into `dst` if the handshake has finished. Otherwise if we're still handshaking we need to take whatever handshake data is made and puts it into `dst`. 

So let's take the handshaking scenario:

 - We'll be ignoring `src` because `PR.Write(...)` will fail until the handshake is complete.
 - So `UpdateHandshakeState()` will write data into the write buffer.
 - We'll drain data from the read buffer and write it into `dst`.

During this:

 - `this_src_write` will be 0 because `PR.Write(...)` returns 0 / `PR_WOULD_BLOCK` until handshake completes.
 - While we're still handshaking in _this_ step, `UpdateHandshakeState()` will write into the read buffer, which then goes into `dst`, incrementing `this_dst_write`.
 - When we run out of handshake data to write, `UpdateHandshakeState()` will quit writing data (and return `PR_WOULD_BLOCK`) and `this_dst_write` will go to data as the write buffer will be empty.

---

A similar thing applies when `src` is non-empty and we're not handshaking:

 - We'll be writing from `src` into the SSL `PRFileDesc *`. `PR.Write(...)` returns number of bytes written which ends up into `this_src_write`. As long as we're still writing data (and not getting an error / `PR_WOULD_BLOCK`), this'll be non-zero. It'll also continue being non-zero while there's space in the write buffer. However, this write buffer gets cleared out by the last step...
 - `UpdateHandshakeState()` will return, a no-op because we're not handshaking.
 - When there's src data and `PR.Write()` succeeds, there'll be data in the write buffer so we'll be able to put it in `dst` and `this_dst_write` will be non-zero. Otherwise, if `src` fails, `this_src_write` becomes 0 and there'll be nothing in the write buffer so `this_dst_write` will go to 0 as well.

---

In all helper methods, we guarantee that on error, 0 is returned instead of a less than zero value.

---

In both cases, when we run out of handshake and/or `src` data, we'll quit writing to the write buffer and `this_dst_write` will eventually go to 0 as well.